File: tests/test_stock_fisico_comprometido.js
'use strict';
/*
 * Desglose de stock (físico / comprometido / devengado) al cotizar o retirar.
 *
 * Daniel, 2026-08-17, con captura del comparador "Asignar y Cotizar" real
 * (dos Gymleco marcados "Sin stock bod. 02" con "1 u. saldo 1" cada uno):
 * "creo que fui yo quien te pidió que solo tomaras el stock teórico [fisico
 * - comprometido - devengado], pero yo requiero mapear la información
 * completa... si hay al menos uno físico positivo y cuántos están
 * comprometidos, para tomar la decisión -- es decir, hay 1 y está
 * comprometido, yo veré si avanzo o no".
 *
 * Antes _tkaStockBadge/_rbaStockBadge colapsaban TODO a un booleano
 * (hay_stock = fisico - comprometido > 0) y mostraban el mismo "Sin stock
 * bod. 02" tanto si el físico era 0 como si era 1 (pero ya comprometido en
 * otro lado) -- exactamente el caso de la captura. Ahora se distinguen dos
 * estados ('sin' | 'comprometido') y el segundo muestra el desglose real a
 * la vista, no solo en el tooltip.
 *
 * Este archivo prueba las DOS copias -- son espejo exacto una de otra,
 * documentado así en el código:
 *   - templates/tickets/_tka_modal.html  (_tkaStockEstado/_tkaStockBadge)
 *   - static/retiros_internal_detail.js  (_rbaStockEstado/_rbaStockBadge)
 *
 * Lo corre tests/test_stock_fisico_comprometido.py (unittest) vía node.
 * Directo:  node tests/test_stock_fisico_comprometido.js
 */

const assert = require('assert');
const fs = require('fs');
const path = require('path');
const vm = require('vm');

const RAIZ = path.resolve(__dirname, '..');

function extraerScript(rutaHtml) {
  const html = fs.readFileSync(path.join(RAIZ, rutaHtml), 'utf8');
  const inicio = html.indexOf('<script>');
  const fin = html.lastIndexOf('</script>');
  assert.ok(inicio !== -1 && fin !== -1, `No encontré <script> en ${rutaHtml}`);
  return html.slice(inicio + '<script>'.length, fin);
}

// ── DOM mínimo (las funciones bajo prueba no tocan el DOM, pero el resto
//    del archivo se define igual y no debe reventar al cargar). ──────────
function nuevoElemento() {
  return {
    value: '', textContent: '', innerHTML: '', style: {}, dataset: {},
    classList: { add() {}, remove() {}, toggle() {} },
    addEventListener() {}, focus() {}, querySelector() { return null; },
    querySelectorAll() { return []; },
  };
}
function nuevoContexto() {
  const document = {
    readyState: 'complete',
    addEventListener() {},
    getElementById() { return nuevoElemento(); },
    querySelector() { return null; },
    querySelectorAll() { return []; },
    createElement() { return nuevoElemento(); },
    body: { appendChild() {} },
  };
  const window = { document, addEventListener() {} };
  return vm.createContext({
    window, document, console, JSON, Math,
    setTimeout() { return 0; }, clearTimeout() {},
    fetch() { return Promise.reject(new Error('sin red en las pruebas')); },
    ilusToast() {}, ilusAlert() {}, ilusConfirm() {}, ilusPrompt() {},
    navigator: { userAgent: 'node' },
    MutationObserver: function () { return { observe() {}, disconnect() {} }; },
    // Solo lo usa retiros_internal_detail.js (inyectado por el template
    // real vía Jinja) -- top-level, así que tiene que existir ANTES de
    // cargar el script o revienta al leerlo. _tka_modal.html no lo usa;
    // no está de más que exista igual.
    RETIROS_DETAIL_DATA: {
      reqId: 1, reqCode: 'RET-TEST-1', customerRut: '11111111-1',
      customerNameHtml: 'Cliente de prueba', customerLabel: 'Cliente de prueba',
      pickupDashboardUrl: '/retiros', shouldMountProposeCal: false, pesoInicial: 0,
    },
  });
}

const ctxTka = nuevoContexto();
vm.runInContext(extraerScript('templates/tickets/_tka_modal.html'), ctxTka, { filename: '_tka_modal.html' });

const ctxRba = nuevoContexto();
vm.runInContext(fs.readFileSync(path.join(RAIZ, 'static', 'retiros_internal_detail.js'), 'utf8'), ctxRba, { filename: 'retiros_internal_detail.js' });

function llamar(ctx, nombre, ...args) {
  ctx.__args = args;
  return vm.runInContext(`${nombre}(...__args)`, ctx);
}

// ── Fixtures de stock (mismo shape que get_erp_stock_by_sku/skus en app.py) ─
function stock({ fisico, comprometido = 0, devengado = 0 }) {
  const disponible = fisico - comprometido;
  return {
    stock: {
      sku: 'X', nombre: 'Producto de prueba',
      fisico, comprometido, devengado, disponible,
      hay_stock: disponible > 0,
    },
  };
}

const CASOS = [
  ['tka', ctxTka, '_tkaStockEstado', '_tkaStockBadge'],
  ['rba', ctxRba, '_rbaStockEstado', '_rbaStockBadge'],
];

let pasadas = 0;
function ok(desc) { pasadas++; console.log(`  ok - ${desc}`); }

for (const [prefijo, ctx, fnEstado, fnBadge] of CASOS) {

  // ════════════════════════════════════════════════════════════════════
  // 1) Stock sano (disponible > 0) -- sin badge, sin estado.
  // ════════════════════════════════════════════════════════════════════
  {
    const l = stock({ fisico: 5, comprometido: 1 });
    assert.strictEqual(llamar(ctx, fnEstado, l), '');
    assert.strictEqual(llamar(ctx, fnBadge, l), '');
    ok(`${prefijo}: stock sano no genera badge ni estado`);
  }

  // ════════════════════════════════════════════════════════════════════
  // 2) Cero físico -- sigue siendo el aviso rojo "Sin stock bod. 02".
  // ════════════════════════════════════════════════════════════════════
  {
    const l = stock({ fisico: 0, comprometido: 0 });
    assert.strictEqual(llamar(ctx, fnEstado, l), 'sin');
    const badge = llamar(ctx, fnBadge, l);
    assert.ok(badge.includes('ln-stock-warn'), 'debe usar la clase de aviso rojo');
    assert.ok(badge.includes('Sin stock bod. 02'), 'debe decir "Sin stock bod. 02"');
    assert.ok(!badge.includes('ln-stock-info'), 'NO debe mezclar con la clase informativa');
    ok(`${prefijo}: 0 físico -> "sin", rojo, "Sin stock bod. 02"`);
  }

  // ════════════════════════════════════════════════════════════════════
  // 3) EL CASO REAL DE DANIEL: 1 físico, 1 comprometido (disponible = 0).
  //    Antes esto cala exactamente en el mismo "Sin stock bod. 02" del
  //    caso anterior -- ya NO debe hacerlo.
  // ════════════════════════════════════════════════════════════════════
  {
    const l = stock({ fisico: 1, comprometido: 1 });
    assert.strictEqual(llamar(ctx, fnEstado, l), 'comprometido');
    const badge = llamar(ctx, fnBadge, l);
    assert.ok(badge.includes('ln-stock-info'), 'debe usar la clase informativa (no la de aviso rojo)');
    assert.ok(!badge.includes('ln-stock-warn'), 'NO debe decir que no hay stock: SÍ hay 1 físico');
    assert.ok(!badge.includes('Sin stock bod. 02'), 'el texto "Sin stock" no debe aparecer en este caso');
    // El desglose real, a la vista (no solo en el tooltip):
    assert.ok(badge.includes('1 físico'), 'el físico tiene que verse en el texto, no solo en el title');
    assert.ok(badge.includes('1 comp.'), 'el comprometido tiene que verse en el texto');
    ok(`${prefijo}: 1 físico + 1 comprometido -> "comprometido", ámbar, con el desglose a la vista`);
  }

  // ════════════════════════════════════════════════════════════════════
  // 4) Físico positivo pero comprometido AÚN MÁS alto (disponible < 0):
  //    misma categoría que el caso 3, no debe caer en "sin".
  // ════════════════════════════════════════════════════════════════════
  {
    const l = stock({ fisico: 2, comprometido: 5 });
    assert.strictEqual(llamar(ctx, fnEstado, l), 'comprometido');
    ok(`${prefijo}: físico 2 / comprometido 5 (disponible negativo) sigue siendo "comprometido", no "sin"`);
  }

  // ════════════════════════════════════════════════════════════════════
  // 5) Sin objeto stock (SKU no encontrado en MAEPR) -- no revienta.
  // ════════════════════════════════════════════════════════════════════
  {
    assert.strictEqual(llamar(ctx, fnEstado, {}), '');
    assert.strictEqual(llamar(ctx, fnEstado, null), '');
    assert.strictEqual(llamar(ctx, fnBadge, {}), '');
    ok(`${prefijo}: sin dato de stock no revienta y no marca nada`);
  }

  // ════════════════════════════════════════════════════════════════════
  // 6) El tooltip conserva el desglose completo, incluido devengado.
  // ════════════════════════════════════════════════════════════════════
  {
    const l = stock({ fisico: 1, comprometido: 1, devengado: 8 });
    const badge = llamar(ctx, fnBadge, l);
    assert.ok(badge.includes('Devengado: 8'), 'el devengado tiene que seguir en el tooltip');
    assert.ok(badge.includes('title='), 'el tooltip completo debe existir');
    ok(`${prefijo}: el tooltip sigue con el desglose completo (incluido devengado)`);
  }
}

console.log(`\n${pasadas} verificaciones OK`);
console.log('stock fisico/comprometido OK');

File: tests/test_stock_fisico_comprometido.py
"""Desglose de stock (físico / comprometido / devengado) al cotizar o retirar.

Daniel, 2026-08-17, con captura real del comparador "Asignar y Cotizar" (dos
Gymleco "1 u. saldo 1" marcados en rojo "Sin stock bod. 02"): "creo que fui
yo quien te pidió que solo tomaras el stock teórico [fisico - comprometido -
devengado], pero yo requiero mapear la información completa... si hay al
menos uno físico positivo y cuántos están comprometidos, para tomar la
decisión -- es decir, hay 1 y está comprometido, yo veré si avanzo o no".

El backend (get_erp_stock_by_sku/get_erp_stock_by_skus, app.py) YA devolvía
el desglose completo desde el 2026-08-03 -- eso no cambió y sus pruebas
viven en tests/test_erp_stock_bodega.py, sin tocar. Lo que cambió es la
PRESENTACIÓN: antes se colapsaba todo a un booleano (hay_stock) y el mismo
"Sin stock bod. 02" en rojo salía tanto si el físico era 0 como si era 1
pero ya comprometido -- dos situaciones muy distintas para quien decide.

Este archivo verifica:
  1. El comportamiento real de _tkaStockEstado/_tkaStockBadge (tickets/
     _tka_modal.html) y su espejo _rbaStockEstado/_rbaStockBadge (retiros/
     internal_detail.js), vía el harness de node (test_stock_fisico_
     comprometido.js) -- mismo patrón que test_shipit_operador_seleccion.py.
  2. Que la SELECCIÓN siga igual de cautelosa que antes en los dos estados
     ('sin' y 'comprometido'): ninguno viene pre-marcado, "seleccionar
     todas" sigue sin arrastrarlos. Ese comportamiento NO cambió a propósito
     -- ver el comentario en ambos archivos fuente.
  3. Que los dos archivos sigan siendo un espejo exacto (mismo criterio,
     mismas clases CSS) -- si se desalinean, el próximo bug será "en
     Retiros sí se ve bien pero en Cotizaciones no".

Si node no está instalado, la parte 1 se salta (no falla); 2 y 3 son texto
puro y corren siempre.

Correr con:  py -m unittest tests.test_stock_fisico_comprometido -v
"""
import os
import shutil
import subprocess
import unittest

RAIZ = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
TPL_TKA = os.path.join(RAIZ, "templates", "tickets", "_tka_modal.html")
JS_RBA = os.path.join(RAIZ, "static", "retiros_internal_detail.js")
CSS_RBA = os.path.join(RAIZ, "static", "retiros_internal_detail.css")

NODE = shutil.which("node")


def _leer(path):
    with open(path, encoding="utf-8") as fh:
        return fh.read()


def _correr_node(script, *args):
    proc = subprocess.run(
        [NODE, os.path.join(RAIZ, "tests", script)] + list(args),
        capture_output=True, text=True, encoding="utf-8", errors="replace",
        cwd=RAIZ, timeout=120,
    )
    if proc.returncode != 0:
        raise AssertionError(
            f"{script} falló (exit {proc.returncode})\n"
            f"--- stdout ---\n{proc.stdout}\n--- stderr ---\n{proc.stderr}"
        )
    return proc.stdout


# ══════════════════════════════════════════════════════════════════════════
#  1) Comportamiento real (harness de node) -- ambos archivos, mismos casos
# ══════════════════════════════════════════════════════════════════════════
@unittest.skipIf(NODE is None, "node no está instalado en este equipo")
class TestComportamientoEnElNavegador(unittest.TestCase):

    def test_los_seis_casos_en_ambos_archivos(self):
        salida = _correr_node("test_stock_fisico_comprometido.js")
        self.assertIn("stock fisico/comprometido OK", salida, salida)
        self.assertIn("12 verificaciones OK", salida, salida)
        # El caso real de Daniel, explícito, para que un cambio futuro que
        # rompa justo ESE caso no pase desapercibido detrás del conteo.
        for marca in (
            'tka: 1 físico + 1 comprometido -> "comprometido", ámbar, con el desglose a la vista',
            'rba: 1 físico + 1 comprometido -> "comprometido", ámbar, con el desglose a la vista',
            'tka: 0 físico -> "sin", rojo, "Sin stock bod. 02"',
            'rba: 0 físico -> "sin", rojo, "Sin stock bod. 02"',
        ):
            self.assertIn(marca, salida, salida)

    def test_el_javascript_de_retiros_compila(self):
        proc = subprocess.run([NODE, "--check", JS_RBA],
                              capture_output=True, text=True, cwd=RAIZ, timeout=60)
        self.assertEqual(0, proc.returncode, proc.stderr)


# ══════════════════════════════════════════════════════════════════════════
#  2) La selección sigue igual de cautelosa (decisión deliberada, no un bug)
# ══════════════════════════════════════════════════════════════════════════
class TestLaSeleccionNoCambio(unittest.TestCase):
    """Daniel pidió VER el desglose para decidir él mismo -- no pidió que el
    sistema pre-seleccione líneas con stock disputado. Cambiar el default de
    selección tiene riesgo de negocio real (auto-incluir en un despacho algo
    que ya está comprometido para otro pedido); el pedido textual fue sobre
    INFORMACIÓN, no sobre automatizar la decisión. Estas pruebas fijan que
    ese comportamiento se mantuvo."""

    @classmethod
    def setUpClass(cls):
        cls.tka = _leer(TPL_TKA)
        cls.rba = _leer(JS_RBA)

    def test_ninguno_de_los_dos_estados_viene_premarcado(self):
        # sinStockBod sigue siendo true para AMBOS estados ('sin' Y
        # 'comprometido') -- solo cambió qué clase visual se le pone a la fila.
        for fuente, nombre in ((self.tka, "_tka_modal.html"), (self.rba, "retiros_internal_detail.js")):
            self.assertIn("const sinStockBod = stockEstado !== '';", fuente,
                          f"{nombre}: sinStockBod debe seguir cubriendo los 2 estados")

    def test_seleccionar_todas_sigue_sin_arrastrar_stock_disputado(self):
        # El guard de "toggle all" compara contra dataset.sinStock, que sigue
        # poblándose desde el mismo sinStockBod de arriba -- no se tocó.
        self.assertIn("if (checked && (isZero || sinStockBod)) return;", self.tka)
        self.assertIn("if (checked && (isZero || sinStockBod)) return;", self.rba)


# ══════════════════════════════════════════════════════════════════════════
#  3) Los dos archivos siguen siendo un espejo exacto
# ══════════════════════════════════════════════════════════════════════════
class TestEspejoEntreCotizacionesYRetiros(unittest.TestCase):
    """Si Cotizaciones y Retiros dejan de compartir el mismo criterio, el
    próximo reporte de Daniel es "en Retiros se ve bien pero en Cotizaciones
    no" -- el mismo tipo de bug que esta feature vino a arreglar."""

    @classmethod
    def setUpClass(cls):
        cls.tka = _leer(TPL_TKA)
        cls.rba = _leer(JS_RBA)
        cls.css_rba = _leer(CSS_RBA)

    def test_ambos_tienen_el_helper_de_estado(self):
        self.assertIn("function _tkaStockEstado(l){", self.tka)
        self.assertIn("function _rbaStockEstado(l){", self.rba)

    def test_ambos_distinguen_sin_de_comprometido(self):
        for fuente in (self.tka, self.rba):
            self.assertIn("return (parseFloat(st.fisico) || 0) <= 0 ? 'sin' : 'comprometido';", fuente)

    def test_ambos_tienen_las_dos_clases_css_de_fila(self):
        self.assertIn(".tka-line.is-no-stock", self.tka)
        self.assertIn(".tka-line.is-stock-comprometido", self.tka)
        self.assertIn(".rba-line.is-no-stock", self.css_rba)
        self.assertIn(".rba-line.is-stock-comprometido", self.css_rba)

    def test_ninguno_perdio_el_badge_rojo_original(self):
        # REGLA #4.2: se AGREGA el estado 'comprometido', no se reemplaza
        # el aviso rojo cuando de verdad no hay nada.
        for fuente in (self.tka, self.rba):
            self.assertIn("ln-stock-warn", fuente)
            self.assertIn("Sin stock bod. 02", fuente)

    def test_el_badge_informativo_muestra_el_desglose_no_solo_el_tooltip(self):
        for fuente in (self.tka, self.rba):
            self.assertIn("ln-stock-info", fuente)
            self.assertIn("' físico'", fuente)
            self.assertIn("' comp.'", fuente)


if __name__ == "__main__":
    unittest.main(verbosity=2)
